File: lua/process.lua
local endline = "\n"
local input = io.open("orbit/output.log", "r")
io.input(input)

local timestamp = io.read()
-- TODO: check for formatting issues the first line should be the time stamp
-- Program should end and notify the user of the issue.

local output = io.open("output.log", "a")
io.output(output)
io.write(timestamp, endline)

-- TODO: Harvest data
local data = {}
for line in io.lines() do
	-- TODO: detect timestamps, we can process more data and associate the time
	local curr = tonumber(line)
	if curr ~= nil then
		table.insert(data, curr)
	end
end
io.close(input)
io.write("Data Size: ", #data, endline)

-- TODO: This Stats stuff could be a module.
-- TODO: need ro calculate the info: average, SD...
local stats = {}
function stats.mean(t)
	local sum = 0
	local count = 0
	for _, v in ipairs(t) do
		if type(v) == "number" then
			sum = sum + v
			count = count + 1
		end
	end
	return (sum / count)
end

function stats.mode(t)
	local counts = {}

	for k, v in pairs(t) do
		if counts[v] == nil then
			counts[v] = 1
		else
			counts[v] = counts[v] + 1
		end
	end

	local biggestCount = 0

	for k, v in pairs(counts) do
		if v > biggestCount then
			biggestCount = v
		end
	end

	local temp = {}

	for k, v in pairs(counts) do
		if v == biggestCount then
			table.insert(temp, k)
		end
	end

	return temp
end

io.write("Mean: ", stats.mean(data), "s", endline)

local modes = stats.mode(data)
for _, n in ipairs(modes) do
	io.write("Mode: ", n, "s", endline)
end

io.write(endline)
io.close(output)

File: python/process.py

from typing import List
from math import sqrt

tab = "\t"
end = "\n"

input = open(file="fastapi/output.log", mode = "r")

timestamp = input.readline()

output = open("output.log", "a")
output.write(timestamp)

# Harvest data, write size to output
data = []
for line in input:
    try:
        data.append(float(line))
    except ValueError:
        print("Error line was not a float")
input.close()
output.write(f"Data Size: {len(data)}\n")

def mean(data: List[float]) -> float:
    sum = 0
    count = 0
    for v in data:
        if type(v) == float:
            sum = sum + v
            count = count + 1
    return (sum / count)

def sd(data: List[float]) -> float:
    m = mean(data)
    sum = 0
    count = 0
    for v in data:
        if type(v) == float:
            vm = v - m
            sum = sum + (vm * vm)
            count = count + 1
    return sqrt(sum / (count -1))

def mode(data: List[float]) -> List[float]:
    counts = {}
    for v in data:
        if v in counts:
            counts[v] = counts[v] + 1
        else:
            counts[v] = 1
    biggest_count = 0
    for c in counts:
        if c > biggest_count:
            biggest_count = c
    output = list()
    for v in counts:
        if v == biggest_count:
            output.append(v)
    return output

# TODO: Median
# TODO: Max
# TODO: Min

output.write(f"Mean:{tab}{mean(data)}{end}")
output.write(f"SD:{tab}{sd(data)}{end}")

modes = mode(data)
for m in modes:
    output.write(f"Mode:{tab}{m}{end}")

output.close()

